internal/provider/client_test.go

```diff
@@ -70,25 +84,201 @@ func TestReadApiEndpoint(t *testing.T) {
+		}
+	}))
+	defer serverWithMissingEDAEndpoint.Close()
+
+	serverWithBadEDAJSON := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case apiEndpoint:
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte(`{"apis":{"gateway": "/api/gateway/", "controller": "/api/controller/", "eda": "/api/eda/"}}`)) //nolint:errcheck
+		case controllerEndpoint:
+			w.Write([]byte(`{"current_version": "/api/controller/v2/"}`)) //nolint:errcheck
+		case edaEndpoint:
+			// Write back intentionally invalid JSON
+			w.Write([]byte(`{`)) //nolint:errcheck
+		default:
+			t.Errorf("Expected to request one of '/api/', '/api/controller/', '/api/eda/', got: %s", r.URL.Path)
+		}
+	}))
+	defer serverWithBadEDAJSON.Close()
+
+	serverWithInvalidEDAURL := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case apiEndpoint:
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte(`{"apis":{"gateway": "/api/gateway/", "controller": "/api/controller/", "eda": "/api/eda/"}}`)) //nolint:errcheck
+		case controllerEndpoint:
+			w.Write([]byte(`{"current_version": "/api/controller/v2/"}`)) //nolint:errcheck
+		case edaEndpoint:
+			// Write an invalid url
+			w.Write([]byte(`{"current_version": "://localhost/api/eda/v1/"}`)) //nolint:errcheck
+		default:
+			t.Errorf("Expected to request one of '/api/', '/api/controller/', '/api/eda/', got: %s", r.URL.Path)
+		}
+	}))
+	defer serverWithInvalidEDAURL.Close()
+
+	testTable := []readApiEndpointTestCase{
+		{
+			name:                   "Bad EDA Endpoint",
+			url:                    serverWithMissingEDAEndpoint.URL,
+			expectedControllerPath: "",
+			expectedEDAPath:        "",
+			diagsShouldHaveErr:     true,
+		},
+		{
+			name:                   "Bad EDA JSON",
+			url:                    serverWithBadEDAJSON.URL,
+			expectedControllerPath: "",
+			expectedEDAPath:        "",
+			diagsShouldHaveErr:     true,
+		},
+		{
+			name:                   "Invalid EDA URL",
+			url:                    serverWithInvalidEDAURL.URL,
+			expectedControllerPath: "",
+			expectedEDAPath:        "",
+			diagsShouldHaveErr:     true,
+		},
 	}
 	for _, tc := range testTable {
-		t.Run(tc.Name, func(t *testing.T) {
-			client, diags := NewClient(tc.URL, &MockAuthenticator{}, true, 0) // readApiEndpoint() is called when creating client
-			assert.Equal(t, false, diags.HasError(), fmt.Sprintf("readApiEndpoint() returns errors (%v)", diags))
-			assert.Equal(t, tc.expectedControllerPath, client.getApiEndpoint())
-			assert.Equal(t, tc.expectedEDAPath, client.getEdaApiEndpoint())
+		t.Run(tc.name, func(t *testing.T) {
+			executeReadApiEndpointTestCase(t, tc)
 		})
 	}
 }
 
+func executeReadApiEndpointTestCase(t *testing.T, tc readApiEndpointTestCase) {
+	// readApiEndpoint() is called when creating client
+	client, diags := NewClient(tc.url, &MockAuthenticator{}, true, 0)
```

Comment:
I see that you moved this check to its own helper function, rather than having it inside the `t.Run()` block. No objection there (since it seems to be de-duplicating common code), but I think when you do this you should call [t.Helper()](https://pkg.go.dev/testing#T.Helper). See also https://dev.to/eminetto/using-test-helpers-in-go-1o55.